contract/src/test/kotlin/io/provenance/scope/loan/contracts/RecordLoanContractUnitTest.kt

```diff
@@ -77,14 +156,118 @@ class RecordLoanContractUnitTest : WordSpec({
+                        assetBuilder.id = randomAssetId
+                        assetBuilder.type = randomNewType
+                        assetBuilder.putKv("loan", unchangedLoan)
+                    }.build()
+                    if (randomExistingType != randomNewType) {
+                        shouldThrow<ContractViolationException> {
+                            RecordLoanContract(
+                                existingAsset = existingAsset,
+                                existingENote = ENote.getDefaultInstance(), // Unused
+                            ).recordAsset(newAsset)
+                        }.let { exception ->
+                            exception.message shouldContainIgnoringCase "Cannot change asset type"
+                        }
+                    }
+                }
+            }
+        }
+        "given an input to an empty scope with more than one type of loan" xshould {
+            "throw an appropriate exception" {
+                // TODO: Implement
+            }
+        }
+        "given an input with more than one type of loan to update an existing asset record" xshould {
+            "throw an appropriate exception" {
+                // TODO: Implement
+            }
+        }
+        "given an input with a loan of a different type than the existing Figure Tech loan" should {
+            "throw an appropriate exception" {
+                checkAll(anyUuid, anyUuid, anyUli) { randomAssetId, randomLoanId, randomUli ->
+                    val existingAsset = Asset.newBuilder().also { assetBuilder ->
+                        assetBuilder.id = randomAssetId // To mark the existing asset as being set
+                        assetBuilder.putKv(
+                            "loan",
+                            FigureTechLoan.newBuilder().also { loanBuilder ->
+                                loanBuilder.id = randomLoanId
+                            }.build().toProtoAny()
+                        )
+                    }.build()
+                    val newAsset = Asset.newBuilder().also { assetBuilder ->
+                        assetBuilder.id = randomAssetId
+                        assetBuilder.putKv(
+                            "loan",
+                            MISMOLoanMetadata.newBuilder().also { loanBuilder ->
+                                loanBuilder.uli = randomUli
+                            }.build().toProtoAny()
+                        )
+                    }.build()
+                    shouldThrow<ContractViolationException> {
+                        RecordLoanContract(
+                            existingAsset = existingAsset,
+                            existingENote = ENote.getDefaultInstance(), // Unused
+                        ).recordAsset(newAsset)
+                    }.let { exception ->
+                        exception.message shouldContain "Could not unpack the input asset's \"loan\" as $FigureTechLoanIdentifier"
```

Comment:
We should maybe add something to the message that comes back from this test case that informs the caller that the existing loan uses the `loan` proto and not the supplied `mismoLoan` proto. IMO the "could not unpack" message makes me think my inputs are malformed, not incorrect to begin with.

Reply:
I pushed an [update](https://github.com/provenance-io/loan-package-contracts/compare/3bc360c36c70d5b58734d407c7c94369379e1675..d23046c2fe7978f17c8af504085262847ee91ce9) which now is able to give a message of the format
```
Expected input asset's "loan" to be a class tech.figure.loan.v1beta1.Loan but was actually a class tech.figure.loan.v1beta1.MISMOLoanMetadata
```
by attempting to re-parse after failure with either of those two clazzes.